### PR TITLE
V050 var keyword removal preparations (End-to-End tests)

### PR DESCRIPTION
**Pull request by christianparpart**
This PR eliminates any occurrences of the "var" keyword in the compiled tests and must be merged *before* #4274.

FYI: the last commit in this PR is special and I'd like to get your feedback on this. I did not find any way to express this in Solidity language syntax, and IIRC, any `FunctionType` not of `Kind::External` or `Kind::Internal` or `Kind::External` is lacking representation in Solidity language wrt. var-keyword removal.

### CODE REVIEW

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 1]
```diff
@@ -8317,26 +8317,6 @@ BOOST_AUTO_TEST_CASE(using_for_by_name)
...
-{
-	char const* sourceCode = R"(
-		library D { struct s { uint a; } function mul(s storage self, uint x) returns (uint) { return self.a *= x; } }
-		contract C {
-			using D for D.s;
-			D.s public x;
-			function f(uint a) returns (uint) {
-				x.a = 6;
-				var g = x.mul;
-				return g({x: a});
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: It seems you can use `return (x.mul)({x: a});` here, which should basically test the same things.

File: test/contracts/AuctionRegistrar.cpp [hunk 2]
```diff
@@ -67,7 +67,7 @@ contract AuctionSystem {
 	function onAuctionEnd(string _name) internal;
 
 	function bid(string _name, address _bidder, uint _value) internal {
-		var auction = m_auctions[_name];
+		Auction auction = m_auctions[_name];
```
Comment (chriseth) [inline test/contracts/AuctionRegistrar.cpp]: Please add `storage` or `memory` while you are at it to all these, i.e. this should be `Action storage auction = m_auctions[_name]`.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 3]
```diff
@@ -1830,22 +1831,23 @@ BOOST_AUTO_TEST_CASE(transfer_ether)
...
-				return x(block.number - 1);
-			}
-		}
-	)";
-	compileAndRun(code, 0, "C");
-	bytes result = callContractFunction("f()");
-	BOOST_REQUIRE_EQUAL(result.size(), 32);
-	BOOST_CHECK(result[0] != 0 || result[1] != 0 || result[2] != 0);
-}
+// TODO(cp): kill test
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: We can do the same here: `return (block.blockhash)(block.number - 1);`

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 4]
```diff
@@ -352,16 +352,14 @@ BOOST_AUTO_TEST_CASE(dynamic_return_types_not_possible)
...
 			function f(uint) public returns (string);
 			function g() public {
-				var x = this.f(2);
+				string x = this.f(2);
 				// we can assign to x but it is not usable.
 				bytes(x).length;
 			}
 		}
 	)";
 	if (dev::test::Options::get().evmVersion() == EVMVersion::homestead())
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp]: Please remove this check.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 5]
```diff
@@ -352,16 +352,14 @@ BOOST_AUTO_TEST_CASE(dynamic_return_types_not_possible)
...
 		CHECK_ERROR(sourceCode, TypeError, "Explicit type conversion not allowed from \"inaccessible dynamic type\" to \"bytes storage pointer\".");
-	else
-		CHECK_WARNING(sourceCode, "Use of the \"var\" keyword is deprecated");
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp]: I think you should still have `CHECK_SUCCESS` or something in this case.

File: test/contracts/AuctionRegistrar.cpp (hunk 2 above)
Comment (christianparpart) [reply test/contracts/AuctionRegistrar.cpp]: What was the `var` keyword doing? defaulting to `storage`, too?

Comment (chriseth) [reply test/contracts/AuctionRegistrar.cpp]: var uses whatever is on the right hand side. If you specify the type, then it defaults to storage.

**Comment by chriseth** [comment]
There is a failing test, you have to remove the component to fix it.

**Comment by christianparpart** [comment]
@chriseth Thanks, I didn't run with `--evm-version homestead` locally. Fixed & only AppVeyor seems to not have finished testing on the CI side yet. (everything else is green)

File: test/libsolidity/SolidityExpressionCompiler.cpp [hunk 6]
```diff
@@ -230,8 +230,8 @@ BOOST_AUTO_TEST_CASE(int_with_szabo_ether_subdenomination)
 {
 	char const* sourceCode = R"(
 		contract test {
-			constructor() {
-				 var x = 1 szabo;
+			function test () {
```
Comment (chriseth) [inline test/libsolidity/SolidityExpressionCompiler.cpp:233]: Not sure if this will break, but let's merge and see.

**Review by chriseth — APPROVED** [review]

File: test/libsolidity/ABIDecoderTests.cpp [hunk 7]
```diff
@@ -374,7 +374,7 @@ BOOST_AUTO_TEST_CASE(decode_function_type_array)
 			}
 			// uses "decode from memory"
 			function test1_dynamic() public returns (uint) {
-				var x = new function() external returns (uint)[](3);
+				function () external returns (uint)[] memory x = new function() external returns (uint)[](4);
```
Comment (axic) [inline test/libsolidity/ABIDecoderTests.cpp:377]: Why was this changed to 4 items?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 8]
```diff
@@ -561,7 +561,8 @@ BOOST_AUTO_TEST_CASE(for_loop)
 		contract test {
 			function f(uint n) returns(uint nfac) {
 				nfac = 1;
-				for (var i = 2; i <= n; i++)
+				uint i;
+				for (i = 2; i <= n; i++)
```
Comment (axic) [inline test/libsolidity/SolidityEndToEndTest.cpp:565]: Hmm, why not define the variable within the for loop?
